### ISSUE TITLE
Suggestions / New Features

### ISSUE THREAD

**Issue by roscopecoltran** [devopsfaith/krakend#7]
Hi guys,

Hope you are all well !

I am giving a quick look at **krakend** and it inspired me the following notes that I wanted to share with you. The idea is to get your point of view/feedback about the relevancy of these suggestions.

Today, I have 3 problems :

1. cannot find an easy to use web admin to test or create pipeline flows like proposed by krakend
2. create resilient and parallel pipelines aggregating several inputs with an high QPS
3. vertical and horizontal scaling of these aggregated micro-services

**About the API Gateway:**
- Add a programmatic load balancer with circuit breakers, ...
  - use cases: real-time configuration reload, fallbacks and pluggable middle-wares
  - projects related:
    - https://github.com/vulcand/vulcand
- Use yaml/toml files and edit them with a webui (eg, [letmegrpc](https://github.com/gogo/letmegrpc/tree/master/form), and the admin rest boilerplate below)

- Managing cascading flows of requests
  - **use cases**: 
    - **general example**, graph-cms:
      - bind several a chain of services bindings for building a page (eg, [link1](https://graphcms.com/)) with fallbacks or circuit breakers
    - **specific example**, computer vision/deep learning:
      - image pre-processing pipeline (eg, cropping, resizing for deep learning)
      - scene/image mining: request in parallel to analyze the same/shared input image to an object detection service (eg [yolo](pjreddie.com/darknet/yolo/)) with an instance matching service (nb. it means you try to detect a known and indexed pictures like this simple web service [pastec](https://github.com/Visu4link/pastec) )
, image mining like running a request to run a semantic service description + ), graph based cms (eg. )
  - **projects related**:
    - https://github.com/flowbase/flowbase
    - https://github.com/jabong/florest-core
    - https://github.com/msgflo
    - https://github.com/cascades-fbp/cascades
    - https://github.com/trustmaster/goflow
    - https://github.com/ryanpeach/goflow

- create stacks of services and their requests pipelines with docker/kubernetes deployement (eg [gateway](https://github.com/nanoscaleio/gateway))

**Front-end/Admin-UI:**
- **Gateway setup with flow based UI**:
  - https://github.com/noflo/noflo-ui (dockerfiles: [link1](https://github.com/rewasa/noflo-server-graph), [link2](https://github.com/mingfang/docker-noflo-ui), [link3](https://github.com/sejnub/docker-noflo-ui))
  - https://github.com/cascades-fbp/flowmaker
- **General administration**
  - https://github.com/vue-bulma/vue-admin ([demo](https://admin.vuebulma.com/#/))
  - https://github.com/marmelab/admin-on-rest ([demo]( https://github.com/marmelab/admin-on-rest-demo))
  - https://github.com/ForestAdmin/forest-movies ([demo](https://app.forestadmin.com/login?livedemo), nice multi-project management with a larger scope of content management)
  - https://github.com/Onefootball/entity-rest-api ([demo](https://github.com/marmelab/ng-admin))

This message is just to share with you how I see a krakend evolving.
Have a nice day ! Viva la vida a barcelona ! :-)

**ps. Is the krakend web form open sourced ?** 

Cheers,
Richard

**Comment by friedcalamari** [comment]
Hello Richard,

First of all thank you very very much for your notes as they provide us valuable insight and interesting topics to work in, I really appreciate the time you spent writing those and all the references to demonstrate your point.

I will start by introducing something that you have not read from anywhere yet, as I am disclosing it in this comment, new news: We are *planning* a SaaS service that will provide the KrakenD ready to consume, at a simple click. I cannot share much details right now but this is the internal roadmap we have set so we are kind of busy putting all the pieces together for this service.

Let me do now some ellaboration around your notes:

> Add a programmatic load balancer with circuit breakers, use cases: real-time configuration reload, fallbacks and pluggable middle-wares.

From the Internet to the KrakenD as per now every implementation will need to configure its own HA Proxy/ELB/whatever. The responsibility of KrakenD is purely the API Gateway. There is, as per today, load balancing to the origin in an automatic fashion, and the only configuration that offers is the number of identical requests you want to send to the backend per request.

The SaaS service on the other hand will of course count with the load balancing capabilities. But for those installing the binary only, today we recommend using HA Proxy in front of the KrakenD. The Circuit Breaker is provided at all times by the service, and is more configurable in the PRO version. This is the balancing from krakend to the backend.

`Internet -> Balancer -> Krakend(s) -> Krakend Balancing -> Backends`

> Managing cascading flows of requests

We will review all these links carefully. Not currently in our roadmap, but I'd like to go through it with the rest of the team.

> create stacks of services and their requests pipelines with docker/kubernetes deployement (eg gateway)

I recall something like this being discussed and not sure if was inside the backlog, but clustering ate much of the planning.

> Front-end/Admin-UI

The flow based UI is super cool... but we are far from an scenario like this right now. Our UI, the KrakenDesigner is an Angular app that, sincerelly, I would blush if I would open source it. There is no javascript/design/html skills in our team and no knowledge at all of Angular. The application works but I can tell you that is not something I would be proud of sharing. In general our UI needs a lot of work. 

In summary, we launched the service last December and we are just starting to walk, I hope we will be running soon.

Thanks you a lot

**Comment by roscopecoltran** [comment]
Hi,

Hope you are all well and thanks for the reply !

Awesome roadmap guys ! 

Last suggestions, it would be awesome to have the gateway to manage backends with the surveyor protocol from nanomsg. ([flow chart](https://fugangqiang.github.io/media/nanomsg/survey.png)). It would allow to send different type of inputs, not only meta data, like a picture to other backends and aggregate a json response with a fixed timeout for respondents.

References:
- https://github.com/go-mangos/mangos/tree/master/examples/survey
- https://github.com/DanielRenne/mangosServer/tree/master/survey
- https://github.com/daniel-j-h/DistributedSearch#usage
- https://daniel-j-h.github.io/post/distributed-search-nanomsg-bond/ (eg. Recursively Building Service Trees)

For the admin ui, I will try to implement https://github.com/qor/admin as it seems the easiest for the moment.

Have a great day !

Cheers,
Richard

**Comment by friedcalamari** [comment]
Richard, thanks again!

Although we don't have the workforce to go through everything, have in mind that we listen all the suggestions. Keep doing it!

As per the admin UI we are not going to use Go at all, our microservices backends are pure go but the UI will be frontend technologies (HTML and JS) feeding from our backend. We can grow the product faster if we keep it this way.

Thanks Richard!

**Comment by alombarte** [comment]
Closing the issue as it is too general and without any derived action.

Feedback noted.

Thanks
